### PR TITLE
[CL]: Add discount rate to balancer pool incentives

### PR DESCRIPTION
**Pull request by AlpinYukseloglu**
Closes: #4837

## What is the purpose of the change

This PR adds a discount rate to balancer pool incentives as a module param. It is currently added to the CL module since it is used entirely internally there and, secondarily, to avoid a minor types package refactor due to import cycles.

## Brief Changelog

- Add `DiscountRate` module param
- Apply discount rate to balancer full range shares

## Testing and Verifying

- Existing `genesis_test.go` and `incentives_test.go` were adapted to test new behavior

## Documentation and Release Note

  - Does this pull request introduce a new feature or user-facing behavior changes? (no)
  - Is a relevant changelog entry added to the `Unreleased` section in `CHANGELOG.md`? (no)
  - How is the feature or change documented? (not documented)

### CODE REVIEW

File: x/concentrated-liquidity/incentives_test.go [hunk 1]
```diff
@@ -194,20 +194,15 @@ func expectedIncentivesFromRate(denom string, rate sdk.Dec, timeElapsed time.Dur
 // Returns value as truncated sdk.Coins as the primary use of this helper is testing higher level incentives functions such as claiming.
 func expectedIncentivesFromUptimeGrowth(uptimeGrowths []sdk.DecCoins, positionShares sdk.Dec, timeInPool time.Duration, multiplier sdk.Int) sdk.Coins {
 	// Sum up rewards from all inputs
-	totalRewards := sdk.DecCoins(nil)
+	totalRewards := sdk.Coins(nil)
 	for uptimeIndex, uptimeGrowth := range uptimeGrowths {
 		if timeInPool >= types.SupportedUptimes[uptimeIndex] {
-			totalRewards = totalRewards.Add(uptimeGrowth...)
+			curRewards := uptimeGrowth.MulDecTruncate(positionShares).MulDecTruncate(multiplier.ToDec())
+			totalRewards = totalRewards.Add(sdk.NormalizeCoins(curRewards)...)
```
Comment (czarcas7ic) [inline x/concentrated-liquidity/incentives_test.go:201]: What is the reason we truncate during every loop iteration instead of truncating at the end?

File: x/concentrated-liquidity/incentives_test.go [hunk 2]
```diff
@@ -4023,6 +4018,10 @@ func (s *KeeperTestSuite) TestClaimAndResetFullRangeBalancerPool() {
 			expectedTokensInGauge := expectedIncentivesFromUptimeGrowth(tc.uptimeGrowth, addedLiquidity, longestSupportedUptime, defaultMultiplier)
 
 			// Ensure gauge coins and amountClaimed are correct
+			fmt.Println("qualifying shares in test: ", addedLiquidity)
+			fmt.Println("expectedTokensInGauge", expectedTokensInGauge)
+			fmt.Println("gauge.Coins", gauge.Coins)
+			fmt.Println("amountClaimed", amountClaimed)
```
Comment (czarcas7ic) [inline x/concentrated-liquidity/incentives_test.go:4024]:
```suggestion
```

File: x/concentrated-liquidity/types/errors.go [hunk 3]
```diff
@@ -576,3 +576,11 @@ type PositionsNotInSameTickRangeError struct {
 func (e PositionsNotInSameTickRangeError) Error() string {
 	return fmt.Sprintf("positions not in same tick range, position 1 tick lower (%d), position 1 tick upper (%d), position 2 tick lower (%d), position 2 tick upper (%d)", e.Position1TickLower, e.Position1TickUpper, e.Position2TickLower, e.Position2TickUpper)
 }
+
+type InvalidDiscountRateError struct {
+	DiscountRate sdk.Dec
+}
+
+func (e InvalidDiscountRateError) Error() string {
+	return fmt.Sprintf("Discount rate must be in range (0, 1]. Attempted to set as %s", e.DiscountRate)
```
Comment (czarcas7ic) [inline x/concentrated-liquidity/types/errors.go:585]: Change error if we decide to change range

File: x/concentrated-liquidity/types/genesis/genesis.go [hunk 4]
```diff
@@ -21,5 +23,8 @@ func (gs GenesisState) Validate() error {
 	if gs.NextPositionId == 0 {
 		return types.InvalidNextPositionIdError{NextPositionId: gs.NextPositionId}
 	}
+	if gs.Params.DiscountRate.LTE(sdk.ZeroDec()) || gs.Params.DiscountRate.GT(sdk.OneDec()) || (gs.Params.DiscountRate == sdk.Dec{}) {
+		return types.InvalidDiscountRateError{DiscountRate: gs.Params.DiscountRate}
+	}
```
Comment (czarcas7ic) [inline x/concentrated-liquidity/types/genesis/genesis.go:28]: This will need to change if we decide to include 1

File: x/concentrated-liquidity/types/params.go [hunk 5]
```diff
@@ -84,3 +93,14 @@ func validateSwapFees(i interface{}) error {
 
 	return nil
 }
+
+// validateDiscountRate validates that the given parameter is a sdk.Dec. Returns error if the parameter is not of the correc type.
```
Comment (czarcas7ic) [inline x/concentrated-liquidity/types/params.go:97]:
```suggestion
// validateDiscountRate validates that the given parameter is a sdk.Dec. Returns error if the parameter is not of the correct type.
```

**Review by czarcas7ic — APPROVED** [review]
Changes look good, just some comments on what we discussed offline!

File: x/concentrated-liquidity/incentives_test.go (hunk 1 above)
Comment (AlpinYukseloglu) [reply x/concentrated-liquidity/incentives_test.go:201]: Each loop corresponds to claiming tokens from a single accumulator, which truncates `DecCoins` to `Coins`. The previous version of this test helper did not match this behavior, which broke rounding-related edge case tests I added in this PR, so this change just aligns its logic with what we expect to see in core logic

File: x/concentrated-liquidity/incentives.go [hunk 6]
```diff
@@ -170,7 +170,18 @@ func (k Keeper) prepareBalancerPoolAsFullRange(ctx sdk.Context, clPoolId uint64)
...
-	qualifyingFullRangeShares := math.GetLiquidityFromAmounts(clPool.GetCurrentSqrtPrice(), types.MinSqrtPrice, types.MaxSqrtPrice, asset0Amount, asset1Amount)
+	qualifyingFullRangeSharesPreDiscount := math.GetLiquidityFromAmounts(clPool.GetCurrentSqrtPrice(), types.MinSqrtPrice, types.MaxSqrtPrice, asset0Amount, asset1Amount)
+
+	// Get discount ratio from governance-set discount rate. Note that the case we check for is technically impossible, but we include
+	// the check as a guardrail anyway. Specifically, we error if the discount ratio is not [0, 1]. Note that this is different from the
+	// discount _rate_, which is [0, 1].
+	balancerSharesDiscountRatio := sdk.OneDec().Sub(k.GetParams(ctx).BalancerSharesRewardDiscount)
+	if !balancerSharesDiscountRatio.GTE(sdk.ZeroDec()) && !balancerSharesDiscountRatio.LTE(sdk.OneDec()) {
+		return 0, sdk.ZeroDec(), types.InvalidDiscountRateError{DiscountRate: k.GetParams(ctx).BalancerSharesRewardDiscount}
+	}
```
Comment (p0mvn) [inline x/concentrated-liquidity/incentives.go:181]: Per-earlier comment, if we validate on the param level, we might catch this before the param is even set

**Review by p0mvn — APPROVED** [review]
